In [ ]:
!pip install tensorflow==2.0.0-alpha0

In [ ]:
import tensorflow as tf
import numpy as np
from tensorflow import keras # from import로 모듈의 일부만 가져오기 (앞에 tf. 안붙여도 됨!)

/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:523: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:524: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:525: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:526: FutureWarning: Passing (type, 1) or 

In [ ]:
# 신경망(모델) 구성
model = keras.Sequential([keras.layers.Dense(units=1, input_shape=[1])]) # 1개의 층(Dense) & 뉴런(unit) & 입력노드(input_shape)

# 학습 과정 설정
# cf. optimizer : loss를 줄여나가기 위해 다른 추측을 하도록 하는 것 (목적함수를 최소로 하는 점을 찾아가도록.. 미분..)
model.compile(optimizer='sgd', loss='mean_squared_error') # 최적화(optimizer) & 손실 함수(loss, 오차 측정)

# sgd (Stochastic Gradient Descent) : 확률적 경사 하강법
# mse (Mean Squared Error) : 평균 오차 제곱 

In [ ]:
# X, Y 데이터 제공 (Y = 2X - 1)
import numpy as np

xs = np.array([-1.0,  0.0, 1.0, 2.0, 3.0, 4.0], dtype=float)
ys = np.array([-3.0, -1.0, 1.0, 3.0, 5.0, 7.0], dtype=float)

In [ ]:
# 훈련(학습) 
model.fit(xs, ys, epochs=500) # Y값에 X값을 맞추는 방법을 알아내도록 모델에 요청 (500회 훈련 루프 돎)

# loss로 추측이 얼마나 잘 되었는지 오차 측정 & optimizer와 데이터로 다른 추측 반복
# convergence : 수렴 (The process of getting very close to the correct answer. An accuracy approaches 100 percent)

Epoch 1/500
6/6 [==============================] - 0s 7ms/sample - loss: 8.7281
Epoch 2/500
6/6 [==============================] - 0s 642us/sample - loss: 8.5534
Epoch 3/500
6/6 [==============================] - 0s 516us/sample - loss: 8.3826
Epoch 4/500
6/6 [==============================] - 0s 518us/sample - loss: 8.2155
Epoch 5/500
6/6 [==============================] - 0s 445us/sample - loss: 8.0522
Epoch 6/500
6/6 [==============================] - 0s 163us/sample - loss: 7.8925
Epoch 7/500
6/6 [==============================] - 0s 259us/sample - loss: 7.7363
Epoch 8/500
6/6 [==============================] - 0s 511us/sample - loss: 7.5837
Epoch 9/500
6/6 [==============================] - 0s 429us/sample - loss: 7.4344
Epoch 10/500
6/6 [==============================] - 0s 384us/sample - loss: 7.2884
Epoch 11/500
6/6 [==============================] - 0s 188us/sample - loss: 7.1456
Epoch 12/500
6/6 [==============================] - 0s 390us/sample - loss: 7.0060
Epoch 13/500
6/

In [ ]:
# 예측
print(model.predict([10.0]))

[[17.158106]]


**[예측 결과 : 19가 아닌 17.158, 18.98...]**
> 19로 딱 떨어지지 않는 이유?
* 1. **아주 적은 데이터를 사용해 훈련했기 때문에**, 모든 X에 대해서 (다른 이외의 데이터들)Y가 2X에서 1을 뺀 값과 같을 것이라는 보장은 없음
* 2. 결과 값은 **확률**을 처리한 것! 따라서 예측 결과를 처리하는 방법을 조정해야 한다.
